pkg/alertmanager/operator.go

```diff
@@ -176,13 +274,142 @@ func (c *Operator) addHandlers() {
+	o, ok := c.getObject(obj)
+	if ok {
+		level.Debug(c.logger).Log("msg", "AlertmanagerConfig added")
+		c.metrics.TriggerByCounter(monitoringv1alpha1.AlertmanagerConfigKind, "add").Inc()
+
+		c.enqueueForNamespace(o.GetNamespace())
+	}
+}
+
+func (c *Operator) handleAlertmanagerConfigUpdate(old, cur interface{}) {
+	if old.(*monitoringv1alpha1.AlertmanagerConfig).ResourceVersion == cur.(*monitoringv1alpha1.AlertmanagerConfig).ResourceVersion {
+		return
+	}
+
+	o, ok := c.getObject(cur)
+	if ok {
+		level.Debug(c.logger).Log("msg", "AlertmanagerConfig updated")
+		c.metrics.TriggerByCounter(monitoringv1alpha1.AlertmanagerConfigKind, "update").Inc()
+
+		c.enqueueForNamespace(o.GetNamespace())
+	}
+}
+
+func (c *Operator) handleAlertmanagerConfigDelete(obj interface{}) {
+	o, ok := c.getObject(obj)
+	if ok {
+		level.Debug(c.logger).Log("msg", "AlertmanagerConfig delete")
+		c.metrics.TriggerByCounter(monitoringv1alpha1.AlertmanagerConfigKind, "delete").Inc()
+
+		c.enqueueForNamespace(o.GetNamespace())
+	}
+}
+
+// TODO: Do we need to enque secrets just for the namespace or in general?
+func (c *Operator) handleSecretDelete(obj interface{}) {
+	o, ok := c.getObject(obj)
+	if ok {
+		level.Debug(c.logger).Log("msg", "Secret deleted")
+		c.metrics.TriggerByCounter("Secret", "delete").Inc()
+
+		c.enqueueForNamespace(o.GetNamespace())
+	}
+}
+
+func (c *Operator) handleSecretUpdate(old, cur interface{}) {
+	if old.(*v1.Secret).ResourceVersion == cur.(*v1.Secret).ResourceVersion {
+		return
+	}
+
+	o, ok := c.getObject(cur)
+	if ok {
+		level.Debug(c.logger).Log("msg", "Secret updated")
+		c.metrics.TriggerByCounter("Secret", "update").Inc()
+
+		c.enqueueForNamespace(o.GetNamespace())
+	}
+}
+
+func (c *Operator) handleSecretAdd(obj interface{}) {
+	o, ok := c.getObject(obj)
+	if ok {
+		level.Debug(c.logger).Log("msg", "Secret added")
+		c.metrics.TriggerByCounter("Secret", "add").Inc()
+
+		c.enqueueForNamespace(o.GetNamespace())
+	}
+}
+
+// enqueueForNamespace enqueues all Alertmanager object keys that belong to the
+// given namespace or select objects in the given namespace.
+func (c *Operator) enqueueForNamespace(nsName string) {
+	nsObject, exists, err := c.nsAlrtCfgInf.GetStore().GetByKey(nsName)
+	if err != nil {
+		level.Error(c.logger).Log(
+			"msg", "get namespace to enqueue Alertmanager instances failed",
+			"err", err,
+		)
+		return
+	}
+	if !exists {
+		level.Error(c.logger).Log(
+			"msg", fmt.Sprintf("get namespace to enqueue Alertmanager instances failed: namespace %q does not exist", nsName),
+		)
+		return
+	}
+	ns := nsObject.(*v1.Namespace)
+
+	objs, err := c.alrtInfs.List(labels.Everything())
+	if err != nil {
+		level.Error(c.logger).Log(
+			"msg", "listing all Alertmanager instances from cache failed",
+			"err", err,
+		)
+		return
+	}
+
+	for _, obj := range objs {
+		// Check for Alertmanager instances in the namespace.
+		am := obj.(*monitoringv1.Alertmanager)
```

Comment:
we still need to enqueue if the object is in the alertmanager's namespace and `alertmanagerConfigNamespaceSelector` is `nil`.

```suggestion
		am := obj.(*monitoringv1.Alertmanager)
		if am.Namespace == nsName {
			c.enqueue(am)
			return
		}
```